### ISSUE TITLE
How do I compile custom xcat dep?

### ISSUE THREAD

**Issue by Michael-1022** [xcat2/xcat-core#7443]
I downloaded the project source code and executed it/ Builddep.sh, but an error occurred:

Error: This script is intended to be used by xCAT development
Error: The GSA directory (/gsa/pokgsa/projects/x/xcat/build/Linux/xcat dep) does not appear to be mounted, cannot continue!

I modified builddep.sh to specify the directory of xcat dep under/root, but a new error occurred:

===>Making sure that the mklocalrepo.sh file contains execute permission
Ls: cannot access/root/xcat core// Xcat dep build/xcat dep/mklocalrepo. sh: No such file or directory
===>-- find not execute, changing+x
Chmod: cannot access'/root/xcat core// Xcat dep build/xcat dep/mklocalrepo. sh ': No such file or directory
===>Checking if 'replacelinks' is in the xcat deps, removing if there
===>Building the tarball at:/root/xcat core// Xcat dep build
===>Creating xcat dep-202405282128. tar. bz2

How do I compile and generate rpm packages related to xcat dep？

**Comment by samveen** [comment]
`builddep.sh` builds dependencies, not the xCAT packages themselves.  Would you update the first post to contain more information about where you are building the packages, and why you need to build them?

Please check `buildlocal.sh` for the script that builds a local copy of the xCAT packages.

**Comment by WZT666-dev** [comment]
> `builddep.sh` builds dependencies, not the xCAT packages themselves. Would you update the first post to contain more information about where you are building the packages, and why you need to build them?
> 
> Please check `buildlocal.sh` for the script that builds a local copy of the xCAT packages.

Thank you for your reply.
I noticed that it has now been updated to xcat2.17.0. I would like to try compiling and generating a new version of xcat installation package myself. I compiled it on the Centos8.5 system

**Comment by Obihoernchen** [comment]
xcat-dep is a little bit harder because the scripts are build to work in an IBM environment only.

But xcat-core you can compile like this:

# Use CentOS/Alma/Rocky 8 container or VM
```
yum -y install git rpm-build createrepo perl-generators perl-Pod-Html perl-Sys-Syslog perl-Digest-MD5 perl-DBI perl-DBD-SQLite perl-JSON perl-Time-HiRes perl-Text-Balanced perl-XML-Parser perl-XML-Simple
umask 0022
./buildcore.sh BUILDALL=1 UP=0 GPGSIGN=0 DEST="$(pwd)/out"
```

**Comment by WZT666-dev** [comment]
> xcat-dep is a little bit harder because the scripts are built to work in an IBM environment only.
> 
> But xcat-core you can compile like this:
> 
> Use CentOS/Alma/Rocky 8 container or VM and run:
> 
> ```
> yum -y install git rpm-build createrepo perl-generators perl-Pod-Html perl-Sys-Syslog perl-Digest-MD5 perl-DBI perl-DBD-SQLite perl-JSON perl-Time-HiRes perl-Text-Balanced perl-XML-Parser perl-XML-Simple
> umask 0022
> ./buildcore.sh BUILDALL=1 UP=0 GPGSIGN=0 DEST="$(pwd)/out"
> ```

After following your instructions, I encountered an error as follows. I noticed that all the rpm packages in the error section need to be compiled in xcat dep. Is it not possible for me to compile and generate them

> [root@t2 xcat-core]# yum install -y xCAT
Last metadata expiration check: 0:00:10 ago on Thu 30 May 2024 05:54:50 PM CST.
Error: 
 Problem: cannot install the best candidate for the job
  - nothing provides syslinux-xcat needed by xCAT-2.17.0-snap202405301752.x86_64
  - nothing provides elilo-xcat needed by xCAT-2.17.0-snap202405301752.x86_64
  - nothing provides perl(Expect) needed by xCAT-2.17.0-snap202405301752.x86_64
  - nothing provides perl-IO-Stty needed by xCAT-2.17.0-snap202405301752.x86_64
  - nothing provides xnba-undi needed by xCAT-2.17.0-snap202405301752.x86_64
  - nothing provides goconserver >= 0.3.3 needed by xCAT-2.17.0-snap202405301752.x86_64
  - nothing provides ipmitool-xcat >= 1.8.17-1 needed by xCAT-2.17.0-snap202405301752.x86_64
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
[root@t2 xcat-core]#

**Comment by samveen** [comment]
Please add xcat-dep repository details into your yum config. that will fix the above issue, given that yum is being used. 

Please visit http://xcat.org/download.html for details.

**Comment by viniciusferrao** [comment]
Hello, I've just sent a message to the user mailing list regarding this issue. I'm also interested in build the `xcat-dep` repository.

Can someone assist with that? I understand that `xcat-dep` has packages that can be easily compiled like `ipmitool` and `syslinux`.

There's a way to remove the IBM requirements? It easier for us to submit new patches if we can properly compile the software in our development machines.

**Comment by viniciusferrao** [comment]
Now with xCAT being sunsetted can you guys upload the required packages to rebuild the `xcat-dep` repository? The files are located on the PokGSA system from IBM.

**Comment by Obihoernchen** [comment]
I'm sorry but we don't have the IBM specific files either. xcat-dep was a mostly a one time manual approach done by IBM without any real CI/CD.
Maybe IBM could upload some of the remaining files somewhere @gurevichmark @besawn ?

**Comment by viniciusferrao** [comment]
> I'm sorry but we don't have the IBM specific files either. xcat-dep was a mostly a one time manual approach done by IBM without any real CI/CD. Maybe IBM could upload some of the remaining files somewhere [@gurevichmark](https://github.com/gurevichmark) [@besawn](https://github.com/besawn) ?

It's time for us to solve that 🤡

**Comment by Obihoernchen** [comment]
We should also get rid of some of the dependencies as well. For instance ipmitool-xcat. Nowadays, I think all the patches are in upstream ipmitool anyways.

**Comment by viniciusferrao** [comment]
We made huge improvements in this matter and today it is possible to rebuild xCAT Dep without much hassle. @dhilst worked on that, maybe we just need a documentation update to close this one. Can you confirm @dhilst?

**Comment by dhilst** [comment]
Yes, long story short there is a mockbuild-all.pl that dispatches to each dependency build script, these are \<dep\>/mockbuild.pl for most dependencies and mockbuild-perl-packages.pl for the perl packages. This will use [mock](https://rpm-software-management.github.io/mock/) behind the scenes. For ubuntu there is a build-debs-all that work similarly, dispatch to \<dep\>/make_deb.sh

These scripts need a proper documentation and better UX though
